src/docmanager/index.ts

```diff
@@ -0,0 +1,958 @@
+   *
+   * @param kernel - An optional kernel name/id to override the default.
+   */
+  open(path: string, widgetName='default', kernel?: IKernelId): Widget {
+    let widget = new Widget();
+    let manager = this._contentsManager;
+    if (widgetName === 'default') {
+      widgetName = this.listWidgetFactories(path)[0];
+    }
+    let mFactoryEx = this._getModelFactoryEx(widgetName);
+    let lang = mFactoryEx.factory.preferredLanguage(path);
+    let id = this._contextManager.findContext(path, mFactoryEx.name);
+    if (id) {
+      this._createWidget(id, widgetName, widget, kernel);
+      return widget;
+    }
+    let model = mFactoryEx.factory.createNew(lang);
+    let opts = mFactoryEx.contentsOptions;
+    manager.get(path, opts).then(contents => {
+      if (contents.format === 'json') {
+        model.fromJSON(contents.content);
+      } else {
+        model.fromString(contents.content);
+      }
+      model.dirty = false;
+      id = this._createContext(path, model, widgetName, contents);
+      this._createWidget(id, widgetName, widget, kernel);
+    });
+    return widget;
+  }
+
+  /**
+   * Create a new file of the given name.
+   *
+   * @param path - The file path to use.
+   *
+   * @param widgetName - The name of the widget factory to use.
+   *
+   * @param kernel - An optional kernel name/id to override the default.
+   */
+  createNew(path: string, widgetName='default', kernel?: IKernelId): Widget {
+    let widget = new Widget();
+    let manager = this._contentsManager;
+    if (widgetName === 'default') {
+      widgetName = this.listWidgetFactories(path)[0];
+    }
+    let mFactoryEx = this._getModelFactoryEx(widgetName);
+    if (!mFactoryEx) {
+      return;
+    }
+    let lang = mFactoryEx.factory.preferredLanguage(path);
+    let model = mFactoryEx.factory.createNew(lang);
+    let opts = utils.copy(mFactoryEx.contentsOptions);
+    if (opts.format === 'json') {
+      opts.content = model.toJSON();
+    } else {
+      opts.content = model.toString();
+    }
+    manager.save(path, opts).then(contents => {
+      let id = this._createContext(path, model, widgetName, contents);
+      this._createWidget(id, widgetName, widget, kernel);
+    });
+    return widget;
+  }
+
+  /**
+   * Get the path given a widget.
+   */
+  getPath(widget: Widget): string {
+    let id = Private.contextProperty.get(widget);
+    return this._contextManager.getPath(id);
+  }
+
+  /**
+   * Clone a widget.
+   *
+   * #### Notes
+   * This will create a new widget with the same model and context
+   * as the existing widget.
+   */
+  clone(widget: Widget): Widget {
+    let parent = new Widget();
+    let id = Private.contextProperty.get(widget);
+    let name = Private.nameProperty.get(widget);
+    this._createWidget(id, name, parent);
+    return parent;
+  }
+
+  /**
+   * Filter messages on the widget.
+   */
+  filterMessage(handler: IMessageHandler, msg: Message): boolean {
```

Comment:
Since we have control over the container widgets that are actually added, why don't we just make an onClose() method for the container widget instead of installing a message filter on our own widget?